### DISCUSSION TITLE
API: folder and hierachy

### DISCUSSION THREAD

**Discussion by newhinton** [nextcloud/bookmarks#2010] [Q&A] [answered]
I am currently extending the android-app by bisada. With the incoming offline mode, i had to implement a database that holds the data locally. While doing so, i once again stumbled upon the peculiar implementation of folders in bookmarks.

As far as i understand, each folder has an id, and only that id is attached to a bookmark.
Each folder also has a parent-id, which allows the hiearachy to be build. 
I can query the api for a specific folder, but i dont get it's children, only the parent-id.
Only the hierachy api results in the full tree.

This leads me to the following:
If i have a bookmark, i can query it's folder. I can get it's parent, but not it's children. I always have to get the full tree, and then store each folder in that tree, alongside it's children in my database. That allows me to easily traverse the tree in both directions. 

Is this recomended? If not, how should i do it?

Secondary, in what cases can a bookmark have more than one folder?

**Comment by marcelklehr** [comment, marked as answer]
Hey :wave: 

I'm happy to see the android app getting some love!

> Secondary, in what cases can a bookmark have more than one folder?

In the web UI, people can copy bookmarks to other folders which will add the same bookmark to the other folder. The reasoning is that every (user,URL) pair should only exist once.

> Only the hierachy api results in the full tree. [...]  I can get it's parent, but not it's children. I always have to get the full tree

You can also query only the immediate children of a folder as well as all descendants of a folder: 

* [GET /public/rest/v2/folder](https://nextcloud-bookmarks.readthedocs.io/en/latest/folder.html#get-full-hierarchy) will give you the folder hierarchy of folder X with Y layers, but no descendant bookmarks
* [GET /public/rest/v2/folder/(int: folder_id)/children](https://nextcloud-bookmarks.readthedocs.io/en/latest/folder.html#get--public-rest-v2-folder-(int-folder_id)-children) This endpoint gives you n layers of bookmarks and folders descendant from folder_id
